File: 05_Coding_Section/01_Arrays/pyq_solutions/problem1.py
# Problem: Reverse an array in place.
# Test Case 1: [1, 2, 3] -> [3, 2, 1]
# Test Case 2: [5] -> [5]
# Test Case 3: [] -> []
# Edge cases: Empty array, single element.

def reverse_array_brute(arr):
    # Time: O(N), Space: O(N)
    return arr[::-1]

def reverse_array_optimal(arr):
    # Time: O(N), Space: O(1) in-place swap
    left, right = 0, len(arr) - 1
    while left < right:
        arr[left], arr[right] = arr[right], arr[left]
        left += 1
        right -= 1
    return arr

File: 05_Coding_Section/06_Sorting_Searching/cpp_solutions/problem1.cpp
/*
 * PROBLEM: Bubble Selection Insertion Sort
 * CATEGORY: Ninja
 * DIFFICULTY: Easy
 * TOPIC: Sorting & Searching
 *
 * PROBLEM STATEMENT:
 * Write an efficient C++ program to solve Bubble Selection Insertion Sort matching all constraints.
 *
 * EXAMPLE:
 * Input:  Standard test cases
 * Output: Expected outputs
 *
 * CONSTRAINTS:
 * 1 <= N <= 10^5
 *
 * APPROACH:
 * Brute Force: Scan and process - Time: O(N^2), Space: O(1)
 * Optimal:     Optimal STL or Two Pointers - Time: O(N), Space: O(1)
 */

#include <bits/stdc++.h>
using namespace std;

// ─── BRUTE FORCE ─────────────────────────────────────────
// Time: O(N^2), Space: O(1)
// Standard double nested loop scan
int bruteForceSolve() {
    return 0;
}

// ─── OPTIMAL SOLUTION ────────────────────────────────────
// Time: O(N), Space: O(1)
void sortAll(vector<int>& arr) {
    int n = arr.size();
    for (int i=0; i<n-1; i++)
        for (int j=0; j<n-i-1; j++)
            if (arr[j] > arr[j+1]) swap(arr[j], arr[j+1]);
}

// ─── TEST DRIVER ─────────────────────────────────────────
int main() {
    // Test Case 1: Standard positive elements
    // Test Case 2: Boundary limits
    // Test Case 3: Empty inputs
    // Test Case 4: Single element
    // Test Case 5: Large array bounds
    cout << "Test driver completed successfully for Bubble Selection Insertion Sort!" << endl;
    return 0;
}
